STM32: Add I2C RTIO support

gautierg-st opened the pull request:
This PR is the continuation of #71961 started by @teburd.
It introduces RTIO support for STM32 I2C v1 and v2.
It works fine with the `sensor_read` and `sensor_get_decoder` functions but there is currently a limitation with the `sensor_stream` function (used for example in `sensor/accel_polling` and `sensor/stream_fifo` samples).

teburd commented:
There's a test against an fram/sram i2c chip that uses rtio and I would recommend testing against that rather than a sensor as its better isolated.

e.g. this fram board https://learn.adafruit.com/adafruit-i2c-fram-breakout

and this test

https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/i2c/i2c_ram

gautierg-st committed:
drivers: i2c: renaming, formatting and clean up

Rename all functions and macros starting by stm32_i2c_* to i2c_stm32_* to
harmonize the driver.
Reformat some indents for better alignment.
Remove i2c_stm32_get_config function declaration from header since it is
not used.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

teburd commented:
Need to pass CI, but I *know* this worked when I wrote it. And it doesn't look too different.

gautierg-st commented:
> Need to pass CI, but I _know_ this worked when I wrote it. And it doesn't look too different.

It is indeed very similar, I merely rebased to integrate latest modifications and did some cosmetic changes.
I am working on adding I2Cv1 support, but that will come as another PR later.

gautierg-st commented:
Split into two commits: first one to move the common functions into a common file, second one to add the RTIO driver.

erwango reviewed (commented):
Mainly questions about code duplication.
Also, what's the plan versus DMA support, should it be kept excluded from RTIO right now ?

teburd commented:
> Mainly questions about code duplication. Also, what's the plan versus DMA support, should it be kept excluded from RTIO right now ?

DMA is optional, the way I did it with the SAM driver is by applying a heuristic to the size f the transfer. I don't recall the details of ST's DMA and I2C peripherals, but there's often a cutoff when it makes sense to pay the cost of programming the DMA or not. E.g. single word size or smaller transfers might fit entirely the in the device FIFO, no need to go program a DMA.

Mostly for Zephyr and RTIO we want to reduce the number of interrupts. Or at the very least avoid the costly indirect interrupts with scheduler overhead if we don't need them.

E.g. what many drivers do today would be maybe be better done if the interrupt handler were the "DIRECT" kind without the scheduling wrapper, only doing scheduling/pm work once a transfer is complete rather than some FIFO wm/refill signaling happening.

gautierg-st commented:
Last commit introduces the support of I2Cv1 also. But this needs further work to factorize some parts of the code, as is, too much is needlessly duplicated.

etienne-lms commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -540,50 +465,6 @@ void i2c_stm32_smbalert_disable(const struct device *dev)
-
-#define STM32_I2C_IRQ_HANDLER_DECL(index)
-#define STM32_I2C_IRQ_HANDLER_FUNCTION(index)
-#define STM32_I2C_IRQ_HANDLER(index)
```
When `CONFIG_I2C_STM32_INTERRUPT` is disabled, **i2c_ll_stm32.c** requires stubbed `STM32_I2C_IRQ_HANDLE*()` macros. This is case sill to be considered (from **i2c_ll_stm32.h**)? I think it is still possible to disable `CONFIG_I2C_STM32_INTERRUPT` from menuconfig, even if quite unlikely used.

gautierg-st replied:
Done

etienne-lms reviewed (commented):
some minor comments.

gautierg-st committed:
drivers: i2c: move functions to a common file

In preparation of the introduction of the STM32 I2Cv2 RTIO driver, move
some functions that are used in both drivers into a common file.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: i2c: stm32 i2cv2 controller driver using rtio

Adds the simplest possible I2Cv2 controller driver for STM32 with the RTIO
interface. Currently only interrupt driven transfers are supported.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>
Co-authored-by: Tom Burdick <thomas.burdick@intel.com>

gautierg-st committed:
drivers: i2c: stm32 i2cv1 controller driver using rtio

Adds the simplest possible I2Cv1 controller driver for STM32 with the RTIO
interface. Currently only interrupt driven transfers are supported.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented on drivers/i2c/i2c_ll_stm32_v2_rtio.c:
```diff
@@ -0,0 +1,514 @@
+#define STM32_I2C_INIT(index)									\
+STM32_I2C_IRQ_HANDLER_DECL(index);								\
+												\
+IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_i2c_v2),						\
```
Moved to common RTIO file

gautierg-st commented on drivers/i2c/i2c_ll_stm32_v2_rtio.c:
```diff
@@ -0,0 +1,514 @@
+												\
+static struct i2c_stm32_data_rtio i2c_stm32_dev_data_##index = {				\
+	.ctx = &CONCAT(_i2c, index, _stm32_rtio),						\
+};												\
```
"Standard" I2C driver has target, DMA, bus recovery and SMBus support (which are not present in the RTIO driver), so I've kept things separate for the moment as a compromise between factorization and code readability.

gautierg-st commented on drivers/i2c/i2c_ll_stm32_v2_rtio.c:
```diff
@@ -0,0 +1,514 @@
+
+	LL_I2C_EnableReset(i2c);
+	LL_I2C_DisableReset(i2c);
+#endif
```
Moved to common RTIO file

gautierg-st commented on drivers/i2c/i2c_ll_stm32_v2_rtio.c:
```diff
@@ -0,0 +1,514 @@
+#include <zephyr/drivers/i2c/rtio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
```
Done

gautierg-st commented on drivers/i2c/i2c_ll_stm32_v2_rtio.c:
```diff
@@ -0,0 +1,514 @@
+
+#define DT_DRV_COMPAT st_stm32_i2c_v2
+
+
```
Done

gautierg-st commented on drivers/i2c/i2c_ll_stm32_v1_rtio.c:
```diff
@@ -0,0 +1,614 @@
+
+}
+
+static inline void handle_sb(const struct device *dev)
```
They were copied from "standard" I2C driver. Removed, the compiler will decide itself if it should be inline or not

etienne-lms commented on drivers/i2c/i2c_ll_stm32_rtio.c:
```diff
@@ -0,0 +1,260 @@
+	ret = i2c_stm32_configure_timing(dev, i2c_clock);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	ret = clock_control_off(clk, (clock_control_subsys_t)&cfg->pclken[0]);
```
Scratches previous `ret` values

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v1_rtio.c:
```diff
@@ -0,0 +1,379 @@
+	if (data->xfer_len == 0 && i2c_rtio_complete(ctx, 0)) {
+		i2c_stm32_start(dev);
+	}
+
```
remove the extra empty line

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v1_rtio.c:
```diff
@@ -0,0 +1,379 @@
+	const struct i2c_stm32_config *cfg = dev->config;
+	struct i2c_stm32_data *data = dev->data;
+	I2C_TypeDef *i2c = cfg->i2c;
+
```
remove the extra empty line

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v1_rtio.c:
```diff
@@ -0,0 +1,379 @@
+	uint8_t slave;
+
+	if (data->xfer_flags & I2C_MSG_ADDR_10_BITS) {
+		slave = (((saddr & 0x0300) >> 7) & 0xFF);
```
can remove outer parentheses

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v1_rtio.c:
```diff
@@ -0,0 +1,379 @@
+		return;
+	}
+	/* according to STM32F1 errata we need to handle these corner cases in
+	 * specific way.
```
```suggestion
	/* According to STM32F1 errata we need to handle these corner cases in
	 * a specific way.
	 *
```

gautierg-st replied:
Done

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v1_rtio.c:
```diff
@@ -0,0 +1,379 @@
+	uint16_t saddr = data->slave_address;
+	uint8_t slave;
+
+	if (data->xfer_flags & I2C_MSG_ADDR_10_BITS) {
```
`if (data->xfer_flags & I2C_MSG_ADDR_10_BITS != 0U)`

Equivalent comments on a few other tests implementation (line 102, 120, 129, 163, 175, ...)

Ditto in **drivers/i2c/i2c_ll_stm32_v2_rtio.c**.

gautierg-st replied:
Fixed

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v1_rtio.c:
```diff
@@ -0,0 +1,379 @@
+	I2C_TypeDef *i2c = cfg->i2c;
+
+	if (data->xfer_len > 0) {
+		switch (data->xfer_len) {
```
Could add case 0 in the switch/case:
```c
	switch (data->xfer_len) {
	case 0:
		if (data->xfer_flags & I2C_MSG_STOP) {
			LL_I2C_GenerateStopCondition(i2c);
		}
		i2c_stm32_master_mode_end(dev);
		break;
	case 1:
		...

gautierg-st replied:
Done

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2_rtio.c:
```diff
@@ -0,0 +1,274 @@
+		}
+
+		/* Receive next byte */
+		if (LL_I2C_IsActiveFlag_RXNE(i2c)) {
```
Deserves an `else` instructions?

gautierg-st replied:
Not present in the standard driver, so I don't think it is necessary
